[NSRPARMA-36] New driver for supporting the Diligent MCC DAQ 1208LS

sneepstefan opened the pull request:
Driver support for the Diligent MCC DAQ device, type 1208LS based on the Linux driver written by Warren Jasper.
https://github.com/wjasper/Linux_Drivers

sneepstefan commented:
Just to be clear this is not my own code, its written by Warren. Just implemented it for use with Odemis. Did a few small adjustments only.

pieleric commented:
Please also include in this PR the actual odemis driver that uses this file.

sneepstefan commented:
I just thought of splitting those due to the amount line of code added to make PR's like this get merged faster. In principle this is stand-alone driver code right?

pieleric commented:
There essentially nothing to review here, because it *should* be just a copy-paste from the original code.
However it's not a complete picture of why you need this change. It's not used any where in Odemis, and cannot be used as-is by any user. That's why to me I think it only makes sense to review as a part of the "pwrmccdaq" driver. I would rather review all at once. (The GUI code can be separate).

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+        try:
+            self.h.open(0x09db, self.productID, serial)
+        except Exception:
+            raise HwError
```
You should raise an *instance* of an exception. Also, it might be handy to know what caused the exception.
```
        except Exception as ex:
            raise HwError("Failed to find USB device {self.productID}, serial {serial}: {ex}")
```

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+        try:
+            value = self.h.read(8, 500)
+        except:
+            print('DIn: error in reading.')
```
This type of except + print is really not helpful. The print will be lost in the darkness of backend stdout, and immediately after `value[0]` will fail as `value` is not defined, and hide the actual error. I'd rather just drop the whole try/except (for all the commands).

sneepstefan replied:
I'd rather not tinker too much on GH code, but I attempted to remove a few.

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+            mode = 1  # differential
+
+        if (channel > 3 and mode == 1):
+            print('AIn: channel out of range for differential mode.')
```
All `print()` should be changed to logging.error(), otherwise we'll never see them.

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+        '''
+
+        if id < 0 or id > 255:
+            print('SetID: id out of range')
```
`raise ValueError(...)`

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+        try:
+            self.h.write([self.SET_ID, id, 0, 0, 0, 0, 0, 0])
+        except:
+            print('Error in writing id')
```
`raise IOError(f'Error in writing id: {ex}')`

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+#! /usr/bin/python3
```
No need for shebang, as it's not an executable. => drop  this line.

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,700 @@
+  pass
+
+
+class SaturationError(Error):
+  ''' Raised when DAQ is saturated '''
+  pass
```
Unused? => drop.

sneepstefan commented:
Hey guys I have updated the driver file _mccdaq.py (it looks new due to complete copy/paste of different file) and added pwrmccdaq.py for the implementation and simulation part of the driver as well.
It is getting really big, I hope you don't mind reviewing it again.
Test cases will follow.

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+import logging
```
author and license is missing

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+    """
+
+    def __init__(self, name: str, role: str, mcc_device: str,
+                 di_channels: Dict[int, Tuple[str, bool]] = None, **kwargs):
```
```suggestion
                 di_channels: Dict[int, Tuple[str, bool]] = {}, **kwargs):
```

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+        # if device and len(device) < 8:
+        #     device = "0" + device
+        self._name = name
+        self._di_channels = di_channels if di_channels else {}
```
```suggestion
        self._di_channels = di_channels
```

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+            setattr(self, va_name, va)  # set the class VA variable name
+            logging.info(f"{va_name} status activated for component {self._name} on channel {channel}")
+
+        if self._di_channels:
```
Should log a warning or HwError if no `_di_channels`? not sure what's the right approach

sneepstefan replied:
I changed the position of the check for _di_channels being not empty so it will cover _channel_vas as well

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+    """
+
+    def __init__(self, name: str, role: str, mcc_device: str, ao_channels: List[int], do_channels: List[int],
+                 spectra, pwr_curve, di_channels: Dict[int, Tuple[str, bool]] = None, **kwargs):
```
```suggestion
                 spectra, pwr_curve, di_channels: Dict[int, Tuple[str, bool]] = {}, **kwargs):
```

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+                raise ValueError(f"Invalid channel value for status {self._di_channels[channel]}")
+
+            # create a VA with False as default
+            va = model.BooleanVA(False, readonly=True)
```
```suggestion
            va = model.BooleanVA(ttl_high, readonly=True)
```
Maybe I am wrong

sneepstefan replied:
I guess its quite confusing but we like the VA default value to start with False. The ttl_high boolean represents the interpretation of the BooleanVA as in which case to set it to true. When ttl_high is true the BooleanVA should still be false when the signal bit is 0. When the bit changes to 1, the ttl signal is high and the BooleanVA set to true. This way the bit can also be 1 and the BooleanVA be false but then ttl_high should be false of course.

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+        :param di_channels (dict -> int, list(str, bool)):
+            the DIO channel used for TTL signal status change through the MCC device.
+            for example di_channels: {2: ["interlockTriggered", False], 7: ["leftSwitch", True]}.
+            key is the channel number(int), value is a list with VA name(str) and tll high is true(bool),
```
```suggestion
            key is the channel number(int), value is a list with VA name(str) and flag stating if ttl high is True(bool),
```

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+            # create a VA with False as default
+            va = model.BooleanVA(False, readonly=True)
+            # append a new dict with the added VA
+            self._channel_vas[channel] = va_name, ttl_high, va
```
This can work, but I think its better to create a channel class something like a `struct`

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+        # initialize values
+        self.productID = 0x0007a  # USB-1208LS
+        # to keep track of the individual bits and values
+        self.port_a_bit_config = {"0": 0, "1": 0, "2": 0, "3": 0, "4": 0, "5": 0, "6": 0, "7": 0}
+        self.port_b_bit_config = {"0": 1, "1": 1, "2": 1, "3": 1, "4": 1, "5": 1, "6": 1, "7": 1}
```
Instead of a dict, just use a list of 8 integers (or bool). They are already accessible as config[0] to config[7].

Alternatively, you could just store a single int, and do the bit arithmetic in the function. ;-)

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+        DIO_port = self._return_port_config(port_number)
+
+        # use a sorted dict to be sure of the key order
+        for bit in list(sorted(DIO_port.keys()))[::-1]:
+            ret_val += str(DIO_port[bit])
```
Assuming it's a list, you can just write:
`return sum(v << i for i, v in enumerate(DIO_port))`

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+            value = 0x3ff
+        if (value < 0):
+            value = 0
+        self.AO_channels[str(channel)] = value
```
Why converting the channel to a string? You can use directly the channel number (int) as key of the dict.

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,175 @@
+CONFIG_MCC = {
+                "name": "Interlock test", "role": "bsd-interlock",
+                "mcc_device": "fake",
+                "di_channels": {2: ["interlockTriggered", False]},
```
Please make it extra complicated by having a second channel on port B. (eg, `9: ["rightFlag", True]`)

sneepstefan replied:
Did that in test_multiple_di_channels()

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,175 @@
+        self.assertFalse(self.mcc_device._status_thread.is_alive())
+
+    def tearDown(self):
+        if self.mcc_device._status_thread:
```
Drop that `if`. It should always be safe to call `.terminate()`. If it's not, the MCCDevice component should be fixed.

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,175 @@
+    def tearDown(self):
+        pass
+
+    def test_power(self):
```
I think you can mostly copy paste the tests from other drivers, like the pwrcomedi_test.py ones.

sneepstefan commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,420 @@
+            with this approach it is actually possible to keep track of customized status changes.
+        """
+        super().__init__(name, role, **kwargs)
+        # force add a trailing 0 or just add a comment to config file?
+        # if device and len(device) < 8:
+        #     device = "0" + device
```
It should have been preceding instead of trailing, changed it in the docstring

nandishjpatel commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,453 @@
+
+@author: Stefan Sneep
+
+Copyright © 2016 Éric Piel, Delmic
```
The copyright name can be yours @sneepstefan ;)

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,453 @@
+            raise ValueError("Value to set is not between 0 and 255")
+
+        DIO_port = self._return_port_config(port_number)
+        in_val = format(value, '#010b')
+        in_val = in_val.replace("0b", "")
+
+        for num, c in enumerate(in_val[::-1]):
+            DIO_port[num] = int(c)
```
Please use bit manipulation instead of strings. Also it should only modify the bits that are not masked in Dconfig.

I think in bitwise operation, you would write `dio_port ^=  ((dio_port ^ value) & dio_config)`. But that would require they are all ints.
Here, you can convert `value` to a list of booleans `list_val`, and then do
```
for i in range(8):
    if dio_config[i]
        dio_port[i] = list_val[i]
```

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,453 @@
+        if channel > 1 or channel < 0:
+            channel = 0
+        # force automatic clipping
+        if (value > 0x3ff):
+            value = 0x3ff
+        if (value < 0):
+            value = 0
```
Clipping can be done in 1 line:
```suggestion
        value  = min(max(0, value), 0x3ff)
```

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,322 @@
+}
+
+
+# django implementation of tag for use as a decorator in unittest (from django.test.utils)
```
"django" ?! It's good that you indicate where this code comes from, but please first explain what this code does.

sneepstefan replied:
Yes good point, at first it seemed quite handy to make a distinction between tests that use the setUp() and others that don't, and still keep them in the same class. Your suggestion below to use a separate private function to create the MCCDevice class fits this scenario better and I can leave out the "django" wrapper.

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,322 @@
+
+
+class TestMCCDevice(unittest.TestCase):
+    def setUp(self):
+        method = getattr(self, self._testMethodName)
+        tags = getattr(method, 'tags', {})
+        if 'skip_setup' in tags:
+            self.mcc_device = None
+            return  # setUp skipped
+
+        self.mcc_device = pwrmccdaq.MCCDevice(**CONFIG_MCC)
+        time.sleep(1)
```
I think this whole thing could be simplified by having just a function called like `_create_device()` which would be called from each function which needs the setup.

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,322 @@
+        for comp_va, key in zip(va_list, MOD_CONFIG_MCC["di_channels"].keys()):
+            self.assertEqual(comp_va.value, MOD_CONFIG_MCC["di_channels"][key][1])
+
+        # set all DI ports to HIGH
+        mcc_device.device.DConfig(0x01, 0xff)
+        mcc_device.device.DConfig(0x04, 0xff)
+        time.sleep(0.15)  # wait a bit to update the port bit status
+
+        # trigger bits 2 and 3 for port A -> binary 0b11110011 -> hex 0x4d
+        mcc_device.device.DOut(0x01, 0xf3)
+
+        # trigger bits 1, 4, 5 and 7 for port B -> binary 0b01001101 -> hex 0x4d
+        mcc_device.device.DOut(0x04, 0x4d)
+        time.sleep(0.15)  # wait a bit to update the port bit status
```
Does this code work on a real device or just the simulator? Is just the simulator, indicate so. 

I am a little bit confused by the usage of DConfig. I thought it's to indicate which port is used for DO or DI. It should be done by the driver, right?

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,322 @@
+            return  # setUp skipped
+
+        self.mcc_device = pwrmccdaq.MCCDeviceLight(**CONFIG_LIGHT_DEVICE)
+        time.sleep(1)
```
Please add a comment why a sleep is needed.

sneepstefan commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,453 @@
+# other board types of this brand. It is mainly used in single-ended mode pinout.
+# Note, with the USB-1208LS board, only 2 analogue channels can be used as output
+# you should use pins 13 and 14 for channels 0->1.
+# For using the DIO ports set the port bit to 1 for digital output and to 0 for
+# digital input. Use pins 21->28 for channels 0->7 and pins 32->39 for channels 8->15.
```
That is true indeed, and if it needs changes it is possible to set it differently in the code. For now we will keep all the bits on 0 when the driver is initiated.

pieleric replied:
After reading the rest of the driver, I think I'm getting it. I'd suggest to clarify it by rephrasing like: "For selecting the direction of the DIO ports, the DConfig register needs to be set to 1 for output and 0 for input, for each bit."

Note that you're not actually doing this in the code! The low-level driver indeed set everything to 0. That means all ports are for input. So the `do_channels` are not going to work. Did you check the code with the real device recently? I guess you'll need to call DConfig with the do_channel bit mask.

pieleric commented on src/odemis/driver/_mccdaq.py:
```diff
@@ -0,0 +1,674 @@
+        self.h.set_nonblocking(1)
+
+        # set default configuration
+        self.DConfig(self.DIO_PORTA, 0x00)  # Port A output
+        self.DConfig(self.DIO_PORTB, 0x00)  # Port B input
```
The comments don't seem to match the code. Both ports are set to "input" (0) at init, right?

sneepstefan commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,453 @@
+# For using the DIO ports set the port bit to 1 for digital output and to 0 for
+# digital input. Use pins 21->28 for channels 0->7 and pins 32->39 for channels 8->15.
+
+from dataclasses import dataclass
```
Thanks for pointing that out. Might we need that on Ubuntu 18.04 systems we can always bump the Python version to 3.7 right?

pieleric replied:
It'd be very really hard. We could always upgrade the computer. That should be much easier!

sneepstefan replied:
Ah that is also an option I guess.
I though when reading this: https://linuxize.com/post/how-to-install-python-3-7-on-ubuntu-18-04/
Upgrading from source would be not so hard.

sneepstefan commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,453 @@
+        if self.productID == 0x0075 and port_number == usb_1208LS.DIO_AUXPORT:
+            bit_mask = ((bit_mask ^ 0xff) & 0xff)
+
+        self.DOut(port_number, bit_mask)
```
Does it matter for the simulator what the real code does in this case? In addition I must admit I am unsure what the bit mask does and what it is for. Your explanation is not clear to me as well.

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,464 @@
+        # initialize values
+        self.productID = 0x0007a  # USB-1208LS
+        # to keep track of the individual bits and values
+        # self.port_a_bit_config = [0, 0, 0, 0, 0, 0, 0, 0]
+        # self.port_b_bit_config = [0, 0, 0, 0, 0, 0, 0, 0]
```
Please drop this.

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,464 @@
+        if value < 0 or value > 255:
+            raise ValueError("Value to set is not between 0 and 255")
+
+        if port_number == usb_1208LS.DIO_PORTA:
+            self.port_a_bit_config = [bool(value & 1 << i) for i in range(8)]
+        else:
+            self.port_b_bit_config = [bool(value & 1 << i) for i in range(8)]
```
DOut and DIn should not write/read on the `config` register. They should affect a different memory. That said, it's a simulator, so in practice, if you write somewhere, no one will ever really know. So you could just do nothing here, and it'd be fine. 
Still, if you want to be able to call DOut and followed by DIn, to simulate a change in the value read by DIn, then this current is fine... but use a separate register from `config`.

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,464 @@
+        if bit < 0 or bit > 7:
+            raise ValueError("Bit value is not between 0 and 7")
+
+        DIO_port = self._return_port_config(port_number)
+
+        return DIO_port[bit]
```
For DBitIn and DBitOut, as it's a simulator, the values read/written are not really meaningful. You could just always return 0 (or 1, whatever is more convenient for now).
What you could also do on these single bit function is to check the config register and warn if it hasn't been set to the right value beforehand (eg, 1 for BitIn and 0 for BitOut).

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,464 @@
+        if port_number == usb_1208LS.DIO_PORTB:
+            return self.port_b_bit_config
+        else:
+            # Port A as default
```
Please explicitly handle port A, and raise an exception for anything else. That will help to detect more errors in the caller.

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,318 @@
+        mcc_device.device.DOut(0x01, 0xf3)
+
+        # trigger bits 1, 4, 5 and 7 for port B -> binary 0b01001101 -> hex 0x4d
+        mcc_device.device.DOut(0x04, 0x4d)
```
Does this code works on the real hardware? I would expect that it doesn't pass, because DOut doesn't even have any effect with Dconfig=0xff. Even if DConfig was temporarily set to 0x00, afterwards, I don't think the value you output on this port is going to be "recorded" when reading again.

You could make this test case skipped if not simulated (something like `if not TEST_NOHW: self.skipTest(...)` ).
Or you could remove completely this part of the test.
=> I'd suggest to

sneepstefan replied:
I have chosen to merge the test cases of the mcc device with the ones for the light (also using a mcc device). This way the test cases can be ran with real hw or a simulator and also interlocks can be tested this way.

pieleric commented on src/odemis/driver/test/pwrmccdaq_test.py:
```diff
@@ -0,0 +1,318 @@
+
+CONFIG_MCC = {
+                "name": "Interlock test", "role": "bsd-interlock",
+                "mcc_device": "fake",
```
Typically we put by default code that test the real hardware, and set the simulator mode when TEST_NOHW env variable == 1. Could you do the same here?

pieleric commented on src/odemis/driver/pwrmccdaq.py:
```diff
@@ -0,0 +1,464 @@
+        """
+        :param mcc_device (str or None): serial number or None (null in yaml file) for auto-detect.
+            the device is considered a USB-powered MCC DAQ 1208LS device. the serial number should
+            contain an added 0 to the beginning of the s/n.
```
Please add that "fake" makes use of a simulator.
